### PR TITLE
paser, executor: add date literal(#3909)

### PR DESCRIPTION
**Pull request by dreamquster**
Fixed the issue #3909

### CODE REVIEW

File: parser/parser.y [hunk 1]
```diff
@@ -2424,7 +2424,7 @@ ReservedKeyword:
 "ADD" | "ALL" | "ALTER" | "ANALYZE" | "AND" | "AS" | "ASC" | "BETWEEN" | "BIGINT"
 | "BINARY" | "BLOB" | "BOTH" | "BY" | "CASCADE" | "CASE" | "CHANGE" | "CHARACTER" | "CHECK" | "COLLATE"
 | "COLUMN" | "CONSTRAINT" | "CONVERT" | "CREATE" | "CROSS" | "CURRENT_DATE" | "CURRENT_TIME"
-| "CURRENT_TIMESTAMP" | "CURRENT_USER" | "DATABASE" | "DATABASES" | "DAY_HOUR" | "DAY_MICROSECOND"
+| "CURRENT_TIMESTAMP" | "CURRENT_USER" | "DATABASE" | "DATE"  | "DATABASES" | "DAY_HOUR" | "DAY_MICROSECOND"
```
Comment (shenli) [inline parser/parser.y]: According to https://dev.mysql.com/doc/refman/5.7/en/keywords.html
Date is a unreserved keyword.

File: parser/parser.y [hunk 2]
```diff
@@ -2943,6 +2943,14 @@ FunctionCallKeyword:
 	{
 		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: $3.([]ast.ExprNode)}
 	}
+|	"DATE" stringLit
```
Comment (XuHuaiyu) [inline parser/parser.y]: We may need a function to check whether stringLit is a valid input.

In MySQL:
``` sql
mysql> select date'2017-99-99';
ERROR 1525 (HY000): Incorrect DATE value: '2017-99-99'

mysql> select date'2017-2-31';
ERROR 1525 (HY000): Incorrect DATE value: '2017-2-31'

mysql> select date'201712-12';
ERROR 1525 (HY000): Incorrect DATE value: '201712-12'

mysql> select date'20171212';
+----------------+
| date'20171212' |
+----------------+
| 2017-12-12     |
+----------------+
1 row in set (0.01 sec)

mysql> select date'2017/12/12';
+------------------+
| date'2017/12/12' |
+------------------+
| 2017-12-12       |
+------------------+
1 row in set (0.00 sec)

mysql> select date'2017/12-12';
+------------------+
| date'2017/12-12' |
+------------------+
| 2017-12-12       |
+------------------+
1 row in set (0.00 sec)
```

File: parser/parser.y [hunk 3]
```diff
@@ -2943,6 +2943,14 @@ FunctionCallKeyword:
...
+	{
+		tp := types.NewFieldType(mysql.TypeString)
```
Comment (XuHuaiyu) [inline parser/parser.y]: The return fieldType should be mysql.TypeDate,
and charset, collate be binary,
flen, and decimal be 10 and 0 respectively.
mysql> select date'2017/12/12';
Field   1:  `date'2017/12/12'`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DATE
Collation:  binary (63)
Length:     10
Max_length: 10
Decimals:   0
Flags:      NOT_NULL BINARY


+------------------+
| date'2017/12/12' |
+------------------+
| 2017-12-12       |
+------------------+
1 row in set (0.00 sec)

File: executor/executor_test.go [hunk 4]
```diff
@@ -1596,6 +1596,21 @@ func (s *testSuite) TestSelectForUpdate(c *C) {
 
 }
 
+func (s *testSuite) TestDateBuiltin(c *C) {
```
Comment (XuHuaiyu) [inline executor/executor_test.go]: Move this to expression/ integration_test.go

**Comment by XuHuaiyu** [comment]
@dreamquster 
hi, any update?

File: parser/parser.y (hunk 1 above)
Comment (dreamquster) [inline parser/parser.y]: Date is keyword in Table 9.2 Keywords and Reserved Words in MySQL 5.7 from that page.

File: executor/executor_test.go (hunk 4 above)
Comment (dreamquster) [inline executor/executor_test.go]: ok,==

File: parser/parser.y (hunk 1 above)
Comment (tiancaiamao) [inline parser/parser.y]: `DATE` is not reserved words. if it's reserved, it should be `DATE[R]` in the table. @dreamquster

**Comment by dreamquster** [comment]
@XuHuaiyu , TestCases is moved to integrate_test.go. The 'data' function will check the valid of input stringLit and convert it to date type.

**Comment by tiancaiamao** [comment]
LGTM

**Comment by hanfei1991** [comment]
@XuHuaiyu PTAL

File: expression/integration_test.go [hunk 5]
```diff
@@ -1753,3 +1753,41 @@ func (s *testIntegrationSuite) TestOtherBuiltin(c *C) {
...
+
+	r = tk.MustQuery("select date'2017/12/12'")
+	r.Check(testkit.Rows("2017-12-12"))
+
+	r = tk.MustQuery("select date'2017/12-12'")
+	r.Check(testkit.Rows("2017-12-12"))
+
+	rs, _ := tk.Exec("select date'2017-99-99';")
+	_, err := tidb.GetRows(rs)
+	c.Assert(err, NotNil)
```
Comment (zz-jason) [inline expression/integration_test.go]: 1. if err is not nil, please add test to check error message
2. add test for zero date, like: "0000-00-00" or "0000-00-00 00:00:00"

File: parser/parser_test.go [hunk 6]
```diff
@@ -527,6 +527,8 @@ func (s *testParserSuite) TestExpression(c *C) {
 		{"select n'string'", true},
 		// for comparison
 		{"select 1 <=> 0, 1 <=> null, 1 = null", true},
+		// for date literal
+		{"select date'1989-09-10'", true},
```
Comment (zz-jason) [inline parser/parser_test.go]: add test for invalid date literals

**Comment by shenli** [comment]
@dreamquster Thanks!
LGTM

**Pull request merged**